caffe2/operators/numpy_batch_matmul_op.h

```diff
@@ -0,0 +1,240 @@
+#define CAFFE2_OPERATORS_NUMPY_BATCH_MATMUL_OP_H
+
+#include <sstream>
+
+#include "caffe2/core/context.h"
+#include "caffe2/core/operator.h"
+#include "caffe2/utils/math.h"
+
+namespace caffe2 {
+
+template <class Context>
+class NumpyBatchMatMulOp final : public Operator<Context> {
+ public:
+  USE_OPERATOR_CONTEXT_FUNCTIONS;
+  NumpyBatchMatMulOp(const OperatorDef& operator_def, Workspace* ws)
+      : Operator<Context>(operator_def, ws),
+        trans_a_(OperatorBase::GetSingleArgument<int>("trans_a", 0)),
+        trans_b_(OperatorBase::GetSingleArgument<int>("trans_b", 0)),
+        use_scratch_(OperatorBase::GetSingleArgument<int>("use_scratch", 0)) {
+    if (use_scratch_) {
+      scratch_ = std::make_shared<Tensor<Context>>();
+    }
+  }
+
+  ~NumpyBatchMatMulOp() {}
+
+  bool RunOnDevice() override {
+    return DispatchHelper<TensorTypes<float>>::call(this, Input(0));
+  }
+
+  template <typename T>
+  bool DoRunWithType() {
+    const auto& A = Input(0);
+    const auto& B = Input(1);
+    auto* Y = Output(0);
+
+    auto ndims_A = A.ndim();
+    auto dims_A = A.dims();
+    auto ndims_B = B.ndim();
+    auto dims_B = B.dims();
+
+    auto* data_A = A.template data<T>();
+    auto* data_B = B.template data<T>();
+
+    auto dimMismatchErrorString = [](size_t dimnum1,
+                                     size_t dim1,
+                                     size_t dimnum2,
+                                     size_t dim2,
+                                     bool trans_a,
+                                     bool trans_b) {
+      std::stringstream ss;
+      ss << "Expected dimension ";
+      ss << dimnum1;
+      ss << " of tensor A with value ";
+      ss << dim1;
+      ss << " to match dimension ";
+      ss << dimnum2;
+      ss << " of tensor B with value ";
+      ss << dim2;
+      ss << ". trans_a = ";
+      ss << trans_a;
+      ss << " trans_b = ";
+      ss << trans_b;
+      return ss.str();
+    };
+
+    if (ndims_A == 1 && ndims_B == 1) {
+      // vector-vector
+      CAFFE_ENFORCE_EQ(
+          dims_A[0],
+          dims_B[0],
+          "Vector-vector product requires each of the vectors to "
+          "be the same size.");
+      Y->Resize(1);
+      math::Dot<T, Context>(
+          dims_A[0], data_A, data_B, Y->template mutable_data<T>(), &context_);
+    } else {
+      bool A_broadcasted = false, B_broadcasted = false;
+      if (ndims_A == 1) {
+        dims_A.insert(dims_A.begin(), 1);
+        ndims_A = 2;
+        A_broadcasted = true;
+      }
+      if (ndims_B == 1) {
+        dims_B.push_back(1);
+        ndims_B = 2;
+        B_broadcasted = true;
+      }
+      // matrix-matrix with batches
+      // [B1..., M, K] * [B2..., K, N] -> [B..., M, N]
+      // In the event that A or B are one-dimensional, the trailing or leading
+      // 1 is not added to the output tensor's size.
+
+      // First step: partition the tensors into inner and outer blocks.
```

Comment:
it's actually not a complete numpy semantics (it also supports '1' dimensions). But agree, let's skip it for now

Reply:
Could you elaborate on this? Not sure what you're referring to